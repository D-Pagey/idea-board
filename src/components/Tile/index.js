import React from 'react';
import {
  string,
  func,
} from 'prop-types';

import {
  TileWrapper,
  TileHeader,
  TitleInput,
  DeleteIcon,
  TileDescription,
  DateStamp,
  CharacterCount,
  SubtleDiv,
} from './styles';

export default function Tile({
  date,
  deleteTile,
  handleChange,
  title,
  description,
  updated,
}) {
  return (
    <TileWrapper>
      <TileHeader>
        <TitleInput
          defaultValue={title}
          onChange={event => handleChange(event)}
          autoFocus
        />

        <DeleteIcon onClick={deleteTile}>delete</DeleteIcon>
      </TileHeader>

      <TileDescription
        defaultValue={description}
        onChange={event => handleChange(event)}
        type="text"
        placeholder="Start noting your idea..."
        maxLength={140}
      />

      <SubtleDiv>
        <DateStamp>
          {updated ? 'Updated' : 'Created'} at {updated || date}
        </DateStamp>

        {description.length > 120
          && <CharacterCount>Characters left = {140 - description.length}</CharacterCount>}
      </SubtleDiv>
    </TileWrapper>
  );
}

Tile.propTypes = {
  date: string.isRequired,
  deleteTile: func.isRequired,
  handleChange: func.isRequired,
  title: string,
  description: string,
  updated: string,
};

Tile.defaultProps = {
  title: '',
  description: '',
  updated: '',
};
