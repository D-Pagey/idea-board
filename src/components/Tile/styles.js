import styled from 'styled-components';

export const TileWrapper = styled.div`
  background-color: white;
  border: 1px solid darkgray;
  border-radius: 6px;
  box-shadow: 0 10px 20px rgba(0,0,0,0.19), 0 6px 6px rgba(0,0,0,0.23);
  display: flex;
  flex-direction: column;
  margin: 15px 10px;
  padding: 4px 7px;
`;

export const TileHeader = styled.div`
  display: flex;
  justify-content: space-between;
  width: 100%;
`;

export const TitleInput = styled.input.attrs({
  type: 'text',
  placeholder: 'Title',
  name: 'title',
})`
  color: black;
  font-size: 1.5rem;
  padding: 2px;
  border-radius: 6px;
  border: 0;
  width: 75%;
`;

export const DeleteIcon = styled.i.attrs({
  className: 'material-icons',
  role: 'button',
  tabIndex: '0',
})`
  position: relative;
  top: 3px;
`;

export const TileDescription = styled.textarea.attrs({
  type: 'text',
  placeholder: 'Start noting your idea...',
  maxLength: 140,
  name: 'description',
})`
  font-family: 'Montserrat';
  font-size: .9rem;
  margin: 4px 0;
  padding: 3px 2px;
`;

export const SubtleDiv = styled.div`
  display: flex;
  justify-content: space-between;
`;

export const DateStamp = styled.p`
  color: darkgray;
  font-size: .7rem;
  margin: 5px 2px 0;
`;

export const CharacterCount = styled.p`
  color: darkgray;
  font-size: .7rem;
  margin: 5px 2px 0;
`;
