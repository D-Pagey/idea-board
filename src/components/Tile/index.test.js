import React from 'react';
import { shallow } from 'enzyme';
import each from 'jest-each';

import {
  TitleInput, TileDescription, DeleteIcon, DateStamp, CharacterCount,
} from './styles';
import Tile from '.';

const props = {
  date: '11:00',
  deleteTile: () => {},
  handleChange: () => {},
};

describe('Tile component', () => {
  it('should render', () => {
    const wrapper = shallow(<Tile {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should handle delete tile', () => {
    const mock = jest.fn();
    const wrapper = shallow(<Tile {...props} deleteTile={mock} />);
    wrapper.find(DeleteIcon).simulate('click');
    expect(mock).toHaveBeenCalled();
  });

  each([TitleInput, TileDescription]).test('should handle change tile', (TargetComponent) => {
    const mock = jest.fn();
    const mockEvent = {};
    const wrapper = shallow(<Tile {...props} handleChange={mock} />);
    wrapper.find(TargetComponent).simulate('change', mockEvent);
    expect(mock).toHaveBeenCalledWith(mockEvent);
  });

  each`
    TargetComponent    | textKey
    ${TitleInput}      | ${'title'} 
    ${TileDescription} | ${'description'}
  `.test('should set correct defaultValues', ({ TargetComponent, textKey }) => {
    const text = 'Some text';
    const newProps = { ...props, [textKey]: text };
    const wrapper = shallow(<Tile {...newProps} />);
    expect(wrapper.find(TargetComponent).prop('defaultValue')).toBe(text);
  });

  it('should have updated date stamp', () => {
    const updated = '11:00';
    const wrapper = shallow(<Tile {...props} updated={updated} />);
    expect(wrapper.find(DateStamp).prop('children').join('')).toBe(`Updated at ${updated}`);
  });

  each`
    characterCount | hasCharacterCount 
    ${119}         | ${false}
    ${120}         | ${false}
    ${121}         | ${true}
  `.test('should render $characterCount character count correctly', ({ characterCount, hasCharacterCount }) => {
    const description = 'x'.repeat(characterCount);
    const wrapper = shallow(<Tile {...props} description={description} />);

    if (hasCharacterCount) {
      expect(wrapper.find(CharacterCount).prop('children').join('')).toBe(`Characters left = ${140 - characterCount}`);
    } else {
      expect(wrapper.find(CharacterCount)).toHaveLength(0);
    }
  });
});
