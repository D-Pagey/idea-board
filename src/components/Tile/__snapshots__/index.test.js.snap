// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tile component should render 1`] = `
<styled.div>
  <styled.div>
    <styled.input
      autoFocus={true}
      defaultValue=""
      onChange={[Function]}
    />
    <styled.i
      onClick={[Function]}
    >
      delete
    </styled.i>
  </styled.div>
  <styled.textarea
    defaultValue=""
    maxLength={140}
    onChange={[Function]}
    placeholder="Start noting your idea..."
    type="text"
  />
  <styled.div>
    <styled.p>
      Created
       at 
      11:00
    </styled.p>
  </styled.div>
</styled.div>
`;
